File: companion-apps/android/app/src/main/java/app/navon/bike/integration/i18n/DistanceFormatter.kt
package app.navon.bike.integration.i18n

import kotlin.math.roundToInt
import kotlin.math.roundToLong

/**
 * Distance formatting helpers — bridges raw meters to the
 * (number, unit) tuple consumed by cue ICU templates and produces
 * ready-to-render UI labels via `units.distance.*` keys.
 */
object DistanceFormatter {
    private const val FT_PER_M = 3.280839895
    private const val MI_PER_M = 0.0006213712
    private const val KM_THRESHOLD_M = 1000.0
    private const val MI_THRESHOLD_M = 1609.0

    fun roundTo10(n: Double): Long = ((n / 10.0).roundToLong()) * 10L

    /** ICU placeholder bundle for a distance voice cue. */
    fun cueValues(meters: Double, mode: DistanceMode): Map<String, Any> = when (mode) {
        DistanceMode.IMPERIAL -> mapOf(
            "distance" to roundTo10(meters * FT_PER_M),
            "distanceUnit" to "feet",
        )
        DistanceMode.METRIC -> {
            if (meters >= KM_THRESHOLD_M) {
                val km = (meters / 100.0).roundToInt() / 10.0 // one decimal
                mapOf(
                    "distance" to km,
                    "distanceUnit" to "kilometers",
                )
            } else {
                mapOf(
                    "distance" to roundTo10(meters),
                    "distanceUnit" to "meters",
                )
            }
        }
    }

    /** UI label for an arbitrary distance in meters. */
    fun label(meters: Double, mode: DistanceMode): String = when (mode) {
        DistanceMode.IMPERIAL -> {
            if (meters >= MI_THRESHOLD_M) {
                val miles = meters * MI_PER_M
                Strings.t(
                    "units.distance.mi",
                    mapOf("distance" to (miles * 10).roundToInt() / 10.0),
                )
            } else {
                Strings.t(
                    "units.distance.ft",
                    mapOf("distance" to (meters * FT_PER_M).roundToInt()),
                )
            }
        }
        DistanceMode.METRIC -> {
            if (meters >= KM_THRESHOLD_M) {
                val km = meters / 1000.0
                Strings.t(
                    "units.distance.km",
                    mapOf("distance" to (km * 10).roundToInt() / 10.0),
                )
            } else {
                Strings.t(
                    "units.distance.m",
                    mapOf("distance" to meters.roundToInt()),
                )
            }
        }
    }
}

File: companion-apps/ios/CompanionApp/Integration/I18n/DistanceFormatter.swift
// Distance formatting helpers — bridges raw meters to the
// (number, unit) tuple consumed by cue ICU templates and produces
// ready-to-render UI labels via `units.distance.*` keys.

import Foundation

enum DistanceFormatter {
    private static let ftPerM: Double = 3.280839895
    private static let miPerM: Double = 0.0006213712
    private static let kmThresholdM: Double = 1000
    private static let miThresholdM: Double = 1609

    static func roundTo10(_ n: Double) -> Int {
        Int((n / 10.0).rounded()) * 10
    }

    static func cueDistanceAndUnit(meters: Double, mode: DistanceMode) -> (distance: Double, unit: String) {
        switch mode {
        case .imperial:
            return (Double(roundTo10(meters * ftPerM)), "feet")
        case .metric:
            if meters >= kmThresholdM {
                let km = (meters / 100.0).rounded() / 10.0
                return (km, "kilometers")
            }
            return (Double(roundTo10(meters)), "meters")
        }
    }

    static func cueValues(meters: Double, mode: DistanceMode) -> [String: MessageValue] {
        let pair = cueDistanceAndUnit(meters: meters, mode: mode)
        return ["distance": .number(pair.distance), "distanceUnit": .string(pair.unit)]
    }

    static func label(meters: Double, mode: DistanceMode) -> String {
        switch mode {
        case .imperial:
            if meters >= miThresholdM {
                let miles = meters * miPerM
                return T.string("units.distance.mi", ["distance": .number((miles * 10).rounded() / 10)])
            }
            return T.string("units.distance.ft", ["distance": .number(Double(Int((meters * ftPerM).rounded())))])
        case .metric:
            if meters >= kmThresholdM {
                let km = meters / 1000.0
                return T.string("units.distance.km", ["distance": .number((km * 10).rounded() / 10)])
            }
            return T.string("units.distance.m", ["distance": .number(Double(Int(meters.rounded())))])
        }
    }
}
